src/main/java/com/faforever/client/login/LoginController.java

```diff
@@ -279,6 +283,17 @@ private Mono<Void> loginWithCode(String code, URI redirectUri, String codeVerifi
     return loginService.login(code, codeVerifier, redirectUri);
   }
 
+  private String getLoginErrorSpecificCause(Throwable throwable) {
+    String cause = throwable.getMessage();
+    if (ERROR_SCOPE_DENIED.matcher(cause).find()) {
+      return "login.scopeDenied";
+    }
+    if (ERROR_NO_CSRF.matcher(cause).find()) {
+      return "login.noCSRF";
+    }
+    return "login.failed";
+  }
+
```

Comment:
The only thing I might say is instead of matching based off the string of the message if instead we just throw a specific type of exception for scope denied and no csrf. Then we just just use pattern matchiing

Reply:
Can you expand a bit? Where should these be throwing from? 

I experimented a bit with exceptions in the OAuthReceivers file

Reply:
This is already in a catch statement so these exception messages are being created in the oauthRetriever so instead of crafting a custom exception message we should instead create an instance of a custom exception type.